### PR TITLE
Issue #135. Support to update complex type property.

### PR DESCRIPTION
**Pull request by xuzhg**
Related #135.
1. Put property
2. Patch property
3. Delete property

### CODE REVIEW

**Comment by vojtechvit** [comment]
Thanks, good work! :+1:

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs [hunk 1]
```diff
@@ -206,29 +201,75 @@ public IHttpActionResult GetOptionalShapesOfCircle(int key)
         }
 
         [HttpPut]
-        [ODataRoute("Windows({key})/CurrentShape")]
-        public IHttpActionResult ReplaceCurrentShape(int key, Shape shape)
+     //   [ODataRoute("Windows({key})/CurrentShape")]
```
Comment (VikingsFan) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs]: remove?

File: OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs [hunk 2]
```diff
@@ -57,28 +70,96 @@ public override string SelectAction(ODataPath odataPath, HttpControllerContext c
             return null;
         }
 
-        private static IEdmProperty GetProperty(ODataPath odataPath, HttpMethod method)
+        [SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity", Justification = "These are simple conversion function and cannot be split up.")]
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs]: warp the line longer than 120

File: OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs [hunk 3]
```diff
@@ -57,28 +70,96 @@ public override string SelectAction(ODataPath odataPath, HttpControllerContext c
...
+        private static IEdmProperty GetProperty(ODataPath odataPath, HttpMethod method, out string prefix, out ComplexCastPathSegment cast)
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs]: warp the line longer than 120

File: OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs [hunk 4]
```diff
@@ -57,28 +70,96 @@ public override string SelectAction(ODataPath odataPath, HttpControllerContext c
...
+            else if (odataPath.PathTemplate == "~/entityset/key/property/complexcast" ||
+                     odataPath.PathTemplate == "~/entityset/key/cast/property/complexcast" ||
+                     odataPath.PathTemplate == "~/singleton/property/complexcast" ||
+                     odataPath.PathTemplate == "~/singleton/cast/property/complexcast")
             {
-                if (odataPath.PathTemplate == "~/entityset/key/property" ||
-                    odataPath.PathTemplate == "~/entityset/key/cast/property" ||
-                    odataPath.PathTemplate == "~/singleton/property" ||
-                    odataPath.PathTemplate == "~/singleton/cast/property")
+                PropertyAccessPathSegment tempSegment = (PropertyAccessPathSegment)odataPath.Segments[odataPath.Segments.Count - 2];
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs]: same here

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 5]
```diff
@@ -473,18 +474,18 @@ public static void UpdateConfiguration(HttpConfiguration configuration)
...
-        [Theory(Skip = "Support deserialize complex type value")]
+        [Theory]
         [InlineData("convention")]
         [InlineData("explicit")]
         // PUT ~/Widnows(1)/CurrentShape
         public async Task PutCurrentShape(string modelMode)
         {
             string serviceRootUri = string.Format("{0}/{1}", BaseAddress, modelMode).ToLower();
-            string requestUri = serviceRootUri + "/Windows(3)/CurrentShape";
+            string requestUri = serviceRootUri + "/Windows(1)/CurrentShape/WebStack.QA.Test.OData.ComplexTypeInheritance.Circle";
```
Comment (VikingsFan) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: //arrange //act //assert

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 6]
```diff
@@ -503,6 +504,59 @@ public static void UpdateConfiguration(HttpConfiguration configuration)
                 String.Format("\nExpected that Radius: 5, but actually: {0},\n request uri: {1},\n response payload: {2}", radius, requestUri, contentOfString));
         }
 
+        [Theory]
+        [InlineData("convention")]
+        [InlineData("explicit")]
+        // PATCH ~/Windows(3)/OptionalShapes
+        public async Task PatchToCollectionComplexTypePropertyNotSupported(string modelMode)
+        {
+            string serviceRootUri = string.Format("{0}/{1}", BaseAddress, modelMode).ToLower();
```
Comment (VikingsFan) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: same here

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 7]
```diff
@@ -503,6 +504,59 @@ public static void UpdateConfiguration(HttpConfiguration configuration)
...
+
+            Assert.True(HttpStatusCode.NotFound == response.StatusCode);
+        }
+
+        [Theory]
+        [InlineData("convention")]
+        [InlineData("explicit")]
+        public async Task PatchToSingleComplexTypeProperty(string modelMode)
+        {
+            string serviceRootUri = string.Format("{0}/{1}", BaseAddress, modelMode).ToLower();
```
Comment (VikingsFan) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: same here

File: OData/test/E2ETest/WebStack.QA.Test.OData/OpenType/TypedTest.cs [hunk 8]
```diff
@@ -399,6 +399,146 @@ public static void UpdateConfiguration(HttpConfiguration configuration)
             }
         }
 
+        [Fact]
+        public async Task PatchOpenComplexTypeProperty()
+        {
+            foreach (string routing in Routings)
+            {
+                await ResetDatasource();
+
```
Comment (VikingsFan) [inline OData/test/E2ETest/WebStack.QA.Test.OData/OpenType/TypedTest.cs]: same here

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs (hunk 1 above)
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs]: Removed.

File: OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs (hunk 2 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs]: Resolved.

File: OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs (hunk 3 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs]: Resolved.

File: OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs (hunk 4 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Routing/Conventions/PropertyRoutingConvention.cs]: Resolved.

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 5 above)
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: Resolved.

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 6 above)
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: Resolved.

File: OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 7 above)
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: resolved.

File: OData/test/E2ETest/WebStack.QA.Test.OData/OpenType/TypedTest.cs (hunk 8 above)
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/OpenType/TypedTest.cs]: It's ok. I test for three scenarios.
Basically, it's not necessary for e2e test cases to add // Arrange // Act // Assert.

**Comment by xuzhg** [comment]
Merged. Thanks.

**Comment by egbertn** [comment]
This unfortunately, is not working, still for OData 7.0
I needed 7.0.0 but cannot go back to 5.9.x (non-Core .net)

**Comment by nidhimj22** [comment]
+1 This doesn't work still.

**Comment by xuzhg** [comment]
Oops. Re-open accident.

**Comment by xuzhg** [comment]
@nidhimj22 @egbertn 

Would you please help file an issue for us to triage?

**Comment by nidhimj22** [comment]
@xuzhg Done! 

https://github.com/OData/WebApi/issues/992

Please have a look. Let me know if you need any other details.

**Pull request closed without merging**